# Task 1

## a) What is the fundamental difference between the model’s parameters and hyperparameters?

Parameters are values that the model learns automatically from the training data during the learning process. Examples of parameters in a neural network are weights and biases. During training, optimization algorithms such as Gradient Descent update these parameters to minimize the loss function and improve the model's predictions.

Hyperparameters are settings that are chosen before training begins. They control how the learning process is performed but are not learned from the data. Examples of hyperparameters include the learning rate, number of hidden layers, number of neurons per layer, batch size, and number of training epochs.

In summary, parameters are learned by the model during training, whereas hyperparameters are specified by the user before training starts.

## b) What is the validation set used for? What would happen if samples from the train/test set were used for the same purpose?

A validation set is used to evaluate the model during training, tune hyperparameters, detect overfitting, and estimate how well the model generalizes to unseen data. It provides an unbiased evaluation during model development without affecting the final test results.

If samples from the training set were used for validation, the evaluation would be overly optimistic because the model has already used those samples to learn and update its parameters. As a result, the validation performance would not accurately reflect the model's ability to generalize to new data.

If samples from the test set were used for validation, information from the test set would influence model selection and hyperparameter tuning. This leads to data leakage and makes the final test evaluation unreliable. Therefore, the test set should only be used after the model and all hyperparameters have been finalized.


# Task 2: Overfitting Analysis

## Scenario

An MLP (Multi-Layer Perceptron) achieves a very low error on the training data but produces a high error on the test data.

---

## What is this phenomenon called?

This phenomenon is called **Overfitting**.

Overfitting occurs when a machine learning model learns the training data too well, including its specific details, noise, and irregular patterns. Instead of learning general features that can be applied to new data, the model effectively memorizes the training examples.

As a result, the model performs very well on the training dataset but poorly on unseen test data because it cannot generalize its knowledge to new samples.

### Example

Suppose an MLP is trained to recognize handwritten digits from the MNIST dataset.

After training:

- Training Accuracy = 99.8%
- Training Error = 0.2%

However, when evaluated on unseen test images:

- Test Accuracy = 80%
- Test Error = 20%

Although the model performs almost perfectly on the training data, it fails to achieve similar performance on new data. This indicates that the model has overfitted the training dataset.

---

## What could cause this problem?

Several factors can lead to overfitting.

### 1. Model Complexity is Too High

If the neural network contains too many hidden layers or too many neurons, it becomes powerful enough to memorize the training data rather than learning general patterns.

#### Example

A problem that can be solved with:

- 2 hidden layers
- 128 neurons per layer

may overfit if a network with:

- 10 hidden layers
- 2000 neurons per layer

is used.

The larger network has significantly more parameters and may simply memorize the training examples.

---

### 2. Insufficient Training Data

When the training dataset is too small, the model has access to only a limited number of examples.

As a result, it may memorize those examples instead of learning the true underlying patterns.

#### Example

Training a digit recognition model using only 100 images is much more likely to cause overfitting than training it using 60,000 images.

---

### 3. Training for Too Many Epochs

If training continues for too long, the model may eventually start fitting noise and small variations in the training data.

Initially, both training and validation performance improve. However, after a certain point, validation performance begins to decrease while training performance continues to improve.

#### Example

| Epoch | Training Accuracy | Validation Accuracy |
| ----- | ----------------- | ------------------- |
| 5     | 90%               | 88%                 |
| 10    | 95%               | 93%                 |
| 20    | 99%               | 94%                 |
| 40    | 99.9%             | 89%                 |

The decrease in validation accuracy indicates overfitting.

---

### 4. Noisy or Incorrect Data

Datasets may contain:

- Incorrect labels
- Outliers
- Measurement errors

A highly flexible neural network may attempt to learn these incorrect patterns.

#### Example

If an image of the digit "3" is accidentally labeled as "8", the model may try to memorize this mistake, reducing its ability to generalize.

---

### 5. Lack of Regularization

Regularization techniques help prevent the model from becoming excessively specialized to the training data.

Without regularization, the model is free to develop very large and highly specific weights, increasing the risk of overfitting.

---

## What methods could you apply to address it?

Several techniques can be used to reduce overfitting and improve generalization.

### 1. Early Stopping

Early stopping monitors the validation loss during training.

Training is stopped when the validation performance no longer improves.

This prevents the model from memorizing the training data.

#### Example

If the validation loss stops improving after epoch 15, training can be terminated instead of continuing to epoch 50.

---

### 2. Dropout

Dropout randomly disables a percentage of neurons during training.

As a result, the network cannot rely too heavily on any particular neuron and is forced to learn more robust features.

#### Example

Using a dropout rate of 0.5 means that approximately 50% of neurons are randomly deactivated during each training step.

---

### 3. L1 and L2 Regularization

Regularization adds a penalty term to the loss function.

This discourages extremely large weight values and encourages simpler models.

Simpler models generally generalize better to unseen data.

---

### 4. Increase the Amount of Training Data

Providing more training examples helps the model learn the true data distribution rather than memorizing individual samples.

A larger dataset usually improves generalization performance.

---

### 5. Data Augmentation

Data augmentation artificially increases the size of the dataset by creating modified versions of existing samples.

Common augmentation techniques include:

- Rotation
- Scaling
- Translation
- Flipping
- Cropping

This exposes the model to more diverse examples.

---

### 6. Reduce Model Complexity

Using fewer layers or fewer neurons decreases the model's capacity to memorize the training data.

A simpler model often generalizes better.

#### Example

Instead of:

- 5 hidden layers
- 512 neurons per layer

use:

- 2 hidden layers
- 128 neurons per layer

if the task does not require a highly complex architecture.

---

## Conclusion

The phenomenon described in this scenario is called **overfitting**. It occurs when a model performs extremely well on training data but poorly on unseen test data because it has learned specific details and noise instead of general patterns.

Common causes include excessive model complexity, insufficient training data, training for too many epochs, noisy data, and lack of regularization. Overfitting can be reduced through techniques such as early stopping, dropout, regularization, data augmentation, increasing the dataset size, and simplifying the model architecture.

These methods improve the model's ability to generalize and achieve better performance on unseen data.


# Task 3: Hyperparameter Optimization using Random Search

Revisit the MLP for the MNIST digits dataset from the previous exercise.

## Objective

Apply a Random Search to find suitable values for the following hyperparameters:

- Number of layers
- Number of nodes per layer
- Learning rate

## Hyperparameter Selection

For each hyperparameter, choose five reasonable candidate values.

### Number of Layers

Choose 5 candidate values.

### Number of Nodes per Layer

Choose 5 candidate values.

### Learning Rate

Choose 5 candidate values.

## Random Search Procedure

For each iteration:

1. Randomly select one value for the number of layers.
2. Randomly select one value for the number of nodes per layer.
3. Randomly select one value for the learning rate.
4. Train the MLP using the selected hyperparameters.
5. Evaluate the model on the validation dataset.

## Number of Iterations

Perform a total of 10 Random Search iterations.

## Validation Set

Split 20% of the original training dataset and use it as a validation set.

## Evaluation

1. Determine which iteration achieves the best performance on the validation dataset.
2. Report the performance of the best model on the test dataset.
3. Compare the obtained result with the model from the previous exercise.

## Optional Task: Early Stopping

Apply Early Stopping during training:

- Monitor the validation loss.
- Stop training if no improvement occurs for 5 consecutive epochs.


# MNIST Training


In [1]:
import tensorflow as tf
import tensorflow_datasets as tfds
from tensorflow.keras.layers import Input, Flatten, Dense
from tensorflow.keras.models import Model
from tensorflow.keras.callbacks import EarlyStopping
import pandas as pd
import random

## Data Processing

The dataset is already loaded and split into a training, validation and testing set.


In [2]:
gpus = tf.config.experimental.list_physical_devices('GPU')
for gpu in gpus:
    tf.config.experimental.set_memory_growth(gpu, True)

# MNIST train has 60k images; we split 80/20
ds_train, ds_val, ds_test = tfds.load(
    "mnist",
    split=["train[:80%]", "train[80%:]", "test"],
    as_supervised=True,
    shuffle_files=True,
    batch_size=32
)

# Normalize datasets
normalize = lambda x, y: (tf.cast(x, tf.float32) / 255.0, tf.cast(y, tf.int32))
ds_train = ds_train.map(normalize)
ds_val   = ds_val.map(normalize)
ds_test  = ds_test.map(normalize)

In [3]:
layer_options = [1, 2, 3, 4, 5]

node_options = [32, 64, 128, 256, 512]

learning_rate_options = [
    0.1,
    0.01,
    0.001,
    0.0005,
    0.0001
]

In [4]:
def build_model(
    num_layers,
    num_nodes,
    learning_rate
):

    input_layer = Input(shape=(28, 28, 1))

    x = Flatten()(input_layer)

    for _ in range(num_layers):
        x = Dense(
            num_nodes,
            activation="relu"
        )(x)

    output_layer = Dense(
        10,
        activation="softmax"
    )(x)

    model = Model(
        inputs=input_layer,
        outputs=output_layer
    )

    optimizer = tf.keras.optimizers.Adam(
        learning_rate=learning_rate
    )

    model.compile(
        optimizer=optimizer,
        loss="sparse_categorical_crossentropy",
        metrics=["accuracy"]
    )

    return model

In [6]:
results = []

best_model = None
best_val_acc = 0

for iteration in range(10):

    num_layers = random.choice(layer_options)

    num_nodes = random.choice(node_options)

    learning_rate = random.choice(
        learning_rate_options
    )

    print(
        f"\nIteration {iteration+1}"
    )

    model = build_model(
        num_layers,
        num_nodes,
        learning_rate
    )

    early_stopping = EarlyStopping(
        monitor="val_loss",
        patience=5,
        restore_best_weights=True
    )

    history = model.fit(
        ds_train,
        validation_data=ds_val,
        epochs=20,
        callbacks=[early_stopping],
        verbose=0
    )

    val_loss, val_acc = model.evaluate(
        ds_val,
        verbose=0
    )

    results.append([
        iteration + 1,
        num_layers,
        num_nodes,
        learning_rate,
        val_acc
    ])

    if val_acc > best_val_acc:

        best_val_acc = val_acc

        best_model = model


Iteration 1

Iteration 2

Iteration 3

Iteration 4

Iteration 5

Iteration 6

Iteration 7

Iteration 8

Iteration 9

Iteration 10


In [7]:
results_df = pd.DataFrame(
    results,
    columns=[
        "Iteration",
        "Layers",
        "Nodes",
        "Learning Rate",
        "Validation Accuracy"
    ]
)

print(results_df)

   Iteration  Layers  Nodes  Learning Rate  Validation Accuracy
0          1       5    128         0.0100             0.948833
1          2       1     64         0.0001             0.960833
2          3       2    512         0.0010             0.972917
3          4       3    128         0.1000             0.300667
4          5       4     64         0.1000             0.198167
5          6       3    256         0.0100             0.954083
6          7       3    256         0.0100             0.959833
7          8       5    256         0.0001             0.970917
8          9       2     32         0.0005             0.958917
9         10       5    128         0.0010             0.974750


In [8]:
test_loss, test_acc = best_model.evaluate(
    ds_test
)

print(
    "\nBest Validation Accuracy:",
    best_val_acc
)

print(
    "Final Test Accuracy:",
    test_acc
)

313/313 ━━━━━━━━━━━━━━━━━━━━ 0s 727us/step - accuracy: 0.9768 - loss: 0.0904

Best Validation Accuracy: 0.9747499823570251
Final Test Accuracy: 0.9768000245094299


# Task 3: Hyperparameter Optimization using Random Search

## Objective

The objective of this task was to optimize the hyperparameters of an MLP trained on the MNIST handwritten digit dataset using Random Search. The hyperparameters selected for optimization were:

- Number of hidden layers
- Number of nodes per layer
- Learning rate

A validation set containing 20% of the original training data was used to evaluate different hyperparameter configurations. The best-performing model on the validation set was then evaluated on the test dataset.

---

## Hyperparameter Search Space

The following candidate values were used during Random Search:

### Number of Hidden Layers

- 1
- 2
- 3
- 4
- 5

### Number of Nodes per Layer

- 32
- 64
- 128
- 256
- 512

### Learning Rate

- 0.1
- 0.01
- 0.001
- 0.0005
- 0.0001

---

## Validation Set

The original MNIST training dataset contains 60,000 images.

The dataset was split as follows:

- Training Set: 48,000 images (80%)
- Validation Set: 12,000 images (20%)
- Test Set: 10,000 images

The validation set was used to compare different hyperparameter combinations and identify the best-performing model.

---

## Random Search Procedure

Ten random hyperparameter combinations were generated.

For each iteration:

1. A random number of layers was selected.
2. A random number of nodes per layer was selected.
3. A random learning rate was selected.
4. The MLP was trained using the selected configuration.
5. Validation accuracy was recorded.
6. The model with the highest validation accuracy was retained.

Early Stopping was applied with a patience value of 5 epochs. Training was stopped whenever the validation loss did not improve for five consecutive epochs.

---

## Results

| Iteration | Layers | Nodes | Learning Rate | Validation Accuracy |
| --------- | ------ | ----- | ------------- | ------------------- |
| 1         | 5      | 128   | 0.01          | 94.88%              |
| 2         | 1      | 64    | 0.0001        | 96.08%              |
| 3         | 2      | 512   | 0.001         | 97.29%              |
| 4         | 3      | 128   | 0.1           | 30.07%              |
| 5         | 4      | 64    | 0.1           | 19.82%              |
| 6         | 3      | 256   | 0.01          | 95.41%              |
| 7         | 3      | 256   | 0.01          | 95.98%              |
| 8         | 5      | 256   | 0.0001        | 97.09%              |
| 9         | 2      | 32    | 0.0005        | 95.89%              |
| 10        | 5      | 128   | 0.001         | 97.48%              |

---

## Best Hyperparameter Configuration

The best-performing configuration was obtained during Iteration 10.

**Best Hyperparameters:**

- Number of Layers: 5
- Nodes per Layer: 128
- Learning Rate: 0.001

**Validation Accuracy:**

97.47%

**Test Accuracy:**

97.68%

---

## Analysis of Results

The results show that the choice of hyperparameters has a significant impact on model performance.

The configurations using a learning rate of 0.1 performed very poorly, achieving validation accuracies of only 30.07% and 19.82%. This indicates that the learning rate was too large, causing unstable training and preventing the model from converging properly.

In contrast, learning rates of 0.001 and 0.0001 produced the best results. These values allowed the optimizer to make smaller and more stable updates to the model parameters.

The best-performing model used five hidden layers with 128 neurons per layer and achieved the highest validation accuracy of 97.47%. This suggests that a deeper network with a moderate number of neurons was able to learn useful representations of the MNIST digits without severe overfitting.

---

## Comparison with the Previous Exercise

In the previous exercise, a fixed MLP architecture consisting of two hidden layers with 128 and 64 neurons was used.

In this task, Random Search explored multiple hyperparameter combinations and automatically selected the configuration that performed best on the validation dataset.

The final optimized model achieved:

- Validation Accuracy: 97.47%
- Test Accuracy: 97.68%

This demonstrates that hyperparameter optimization can improve model selection and help identify configurations that generalize well to unseen data.

---

## Conclusion

Random Search was successfully applied to optimize the hyperparameters of an MLP trained on the MNIST dataset.

Ten different hyperparameter combinations were evaluated using a validation set. The best model used five hidden layers, 128 neurons per layer, and a learning rate of 0.001. This configuration achieved a validation accuracy of 97.47% and a test accuracy of 97.68%.

The experiment demonstrates that hyperparameter tuning is an important step in deep learning because different hyperparameter choices can significantly influence model performance and generalization capability.


# Task 4: Optimization Challenges

## Question

Inspect the provided error/cost-parameter curve. Identify regions that represent common optimization challenges encountered during backpropagation training. For each marked region:

1. Name the issue.
2. Explain why it occurs.
3. Describe its effect on convergence.

---

# Introduction

Training a neural network involves finding the set of parameters (weights and biases) that minimizes the loss function. Gradient Descent and Backpropagation are used to iteratively update these parameters and reduce the error.

In practice, the loss surface of a neural network is highly complex and contains valleys, flat regions, and irregular shapes. These characteristics can create optimization challenges that affect the speed and quality of convergence. The provided cost-parameter curve illustrates several of these challenges.

---

# 1. Local Minimum

## Description

A local minimum is a point where the cost is lower than in the surrounding region but is not the lowest point on the entire loss surface.

In the given curve, the first shallow valley can be identified as a local minimum.

## Why Does It Occur?

Neural network loss functions are non-convex and often contain multiple valleys. Since Gradient Descent relies only on local gradient information, it may converge to the nearest minimum rather than the best possible solution.

## Effect on Convergence

The optimizer may stop improving once it reaches the local minimum. Although convergence is achieved, the resulting solution is suboptimal because a lower cost value exists elsewhere on the loss surface.

---

# 2. Global Minimum

## Description

The global minimum is the point with the lowest cost value on the entire loss surface.

In the provided graph, the deepest valley represents the global minimum.

## Why Does It Occur?

The global minimum corresponds to the parameter values that minimize the training loss as much as possible.

## Effect on Convergence

This is the desired destination of the optimization process because it represents the best solution available on the loss surface and produces the lowest training error.

---

# 3. Plateau

## Description

A plateau is a large flat region where the cost changes very little even when the parameters are updated.

The extended flat section on the right side of the graph represents a plateau.

## Why Does It Occur?

In plateau regions, the gradient becomes very small. Since Gradient Descent updates parameters based on the gradient, only very small parameter changes are made.

## Effect on Convergence

Learning becomes extremely slow because the optimizer receives little information about the direction in which it should move. As a result, training may require many additional iterations before noticeable improvement occurs.

---

# 4. Saddle Point

## Description

A saddle point is a point where the gradient is close to zero, but the point is neither a minimum nor a maximum.

Although not clearly visible in a simple two-dimensional plot, saddle points frequently occur in the high-dimensional loss surfaces of deep neural networks.

## Why Does It Occur?

Deep neural networks contain a very large number of parameters, creating complex optimization landscapes with many saddle-shaped regions.

## Effect on Convergence

The optimizer may spend a long time near a saddle point because the gradient is very small. This slows down convergence and can make training appear stalled.

---

# 5. Vanishing Gradient

## Description

Vanishing gradients occur when gradients become extremely small during backpropagation.

## Why Does It Occur?

During backpropagation, gradients are repeatedly multiplied as they move through the layers of the network. In deep networks, especially when using activation functions such as Sigmoid or Tanh, these repeated multiplications can cause gradients to shrink toward zero.

## Effect on Convergence

Earlier layers receive almost no learning signal and therefore update their weights very slowly. This makes training difficult and significantly slows convergence.

---

# 6. Exploding Gradient

## Description

Exploding gradients occur when gradients become extremely large during backpropagation.

## Why Does It Occur?

Repeated multiplication of large values across many layers can cause gradients to grow uncontrollably.

## Effect on Convergence

Parameter updates become excessively large, making training unstable. The loss may oscillate, diverge, or become undefined. In severe cases, training can completely fail.

---

# 7. Oscillations

## Description

Oscillations occur when the optimizer repeatedly overshoots the minimum and moves back and forth around it.

## Why Does It Occur?

This usually happens when the learning rate is too high. Large updates cause the optimizer to jump across the minimum instead of gradually approaching it.

## Effect on Convergence

Convergence becomes slow and unstable because the optimizer cannot settle at the optimal solution. In extreme cases, the model may never converge.

---

# 8. Slow Convergence

## Description

Slow convergence refers to situations where the loss decreases very slowly during training.

## Why Does It Occur?

Common causes include:

- Plateau regions
- Very small gradients
- Poor learning rate selection
- Saddle points

## Effect on Convergence

Training requires many additional epochs and computational resources before reaching an acceptable solution.

---

# Summary Table

| Optimization Challenge | Why It Occurs                            | Effect on Convergence                 |
| ---------------------- | ---------------------------------------- | ------------------------------------- |
| Local Minimum          | Multiple valleys in the loss surface     | Converges to a suboptimal solution    |
| Global Minimum         | Lowest point on the loss surface         | Best possible convergence             |
| Plateau                | Very small gradients                     | Slow learning and slow convergence    |
| Saddle Point           | Gradient close to zero but not a minimum | Training slows significantly          |
| Vanishing Gradient     | Gradients shrink during backpropagation  | Early layers learn very slowly        |
| Exploding Gradient     | Gradients grow excessively large         | Unstable training and divergence      |
| Oscillations           | Learning rate is too high                | Overshooting and unstable convergence |
| Slow Convergence       | Small gradients or poor hyperparameters  | Increased training time               |

---

# Conclusion

The cost surface of a neural network contains several optimization challenges that can affect Gradient Descent during training. Important examples include local minima, saddle points, plateaus, vanishing gradients, exploding gradients, oscillations, and slow convergence.

These challenges can slow learning, reduce model performance, or prevent the optimizer from reaching the global minimum. Modern optimization methods such as Adam, RMSProp, Momentum, appropriate learning rate selection, and proper weight initialization are commonly used to reduce these problems and improve convergence during neural network training.
